In [1]:
import os, json, gzip, pathlib
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from tqdm import tqdm
from sklearn.model_selection import train_test_split 
from sklearn.ensemble import RandomForestClassifier
from sklearn.cluster import KMeans
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import classification_report
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC




In [2]:
with open("../dataset/cyberlab_2020-02-29.json") as f:
    data = json.load(f)

In [3]:
events = []
for session in data:
    (session_id, session_events), = session.items()
    for e in session_events:
        e["session_id"] = session_id
        events.append(e)

df = pd.DataFrame(events)
print(df.shape)
df.head()

(379033, 36)


session_id                                dst_host_identifier  \
0  0258c1b64b66  3c84036c67663a493ac99f838e65b00f36c939d90c7814...   
1  0258c1b64b66  3c84036c67663a493ac99f838e65b00f36c939d90c7814...   
2  0258c1b64b66  3c84036c67663a493ac99f838e65b00f36c939d90c7814...   
3  0258c1b64b66  3c84036c67663a493ac99f838e65b00f36c939d90c7814...   
4  0258c1b64b66  3c84036c67663a493ac99f838e65b00f36c939d90c7814...   

                  eventid                    timestamp  \
0  cowrie.session.connect  2020-02-29T00:06:51.433806Z   
1   cowrie.client.version  2020-02-29T00:06:51.457121Z   
2       cowrie.client.kex  2020-02-29T00:06:51.478455Z   
3     cowrie.login.failed  2020-02-29T00:06:51.663160Z   
4   cowrie.session.closed  2020-02-29T00:06:52.686702Z   

                                   src_ip_identifier dst_ip_identifier  \
0  f0bb16e4a6ed33e675e3fdea10ee11afc7f5a9bc57406c...              None   
1  f0bb16e4a6ed33e675e3fdea10ee11afc7f5a9bc57406c...              None   
2  f0bb16e4a6ed33e675e3fdea10ee11afc7f5a9bc57406c...              None   
3  f0bb16e4a6ed33e675e3fdea10ee11afc7f5a9bc57406c...              None   
4  f0bb16e4a6ed33e675e3fdea10ee11afc7f5a9bc57406c...              None   

                                             message protocol  src_port  \
0  New connection: f0bb16e4a6ed33e675e3fdea10ee11...      ssh   53566.0   
1        Remote SSH version: b'SSH-2.0-libssh-0.6.3'     None       NaN   
2  SSH client hassh fingerprint: 51cba57125523ce4...     None       NaN   
3                 login attempt [root/147258] failed     None       NaN   
4                    Connection lost after 1 seconds     None       NaN   

              sensor  ...  type outfile  destfile duplicate shasum   url  \
0  ubuntu_basic_pool  ...  None    None      None      None   None  None   
1  ubuntu_basic_pool  ...  None    None      None      None   None  None   
2  ubuntu_basic_pool  ...  None    None      None      None   None  None   
3  ubuntu_basic_pool  ...  None    None      None      None   None  None   
4  ubuntu_basic_pool  ...  None    None      None      None   None  None   

  ttylog size filename  data  
0   None  NaN     None  None  
1   None  NaN     None  None  
2   None  NaN     None  None  
3   None  NaN     None  None  
4   None  NaN     None  None  

[5 rows x 36 columns]

In [4]:
df.describe(include="all")

session_id                                dst_host_identifier  \
count         379033                                             379033   
unique         71172                                                 24   
top     17ad4067463e  912f523c81eaa22b69e4e40dcd38ff6a50e50a45d2cb13...   
freq              63                                              32697   
mean             NaN                                                NaN   
std              NaN                                                NaN   
min              NaN                                                NaN   
25%              NaN                                                NaN   
50%              NaN                                                NaN   
75%              NaN                                                NaN   
max              NaN                                                NaN   

                    eventid                    timestamp  \
count                379033                       379033   
unique                   10                       379033   
top     cowrie.login.failed  2020-02-29T00:06:51.433806Z   
freq                  80970                            1   
mean                    NaN                          NaN   
std                     NaN                          NaN   
min                     NaN                          NaN   
25%                     NaN                          NaN   
50%                     NaN                          NaN   
75%                     NaN                          NaN   
max                     NaN                          NaN   

                                        src_ip_identifier dst_ip_identifier  \
count                                              379033                 0   
unique                                               2783                 0   
top     374d2a0a69375cd160fba8a3acd47c6b88c8f819f65405...               NaN   
freq                                                17868               NaN   
mean                                                  NaN               NaN   
std                                                   NaN               NaN   
min                                                   NaN               NaN   
25%                                                   NaN               NaN   
50%                                                   NaN               NaN   
75%                                                   NaN               NaN   
max                                                   NaN               NaN   

                                            message protocol      src_port  \
count                                        379033    71172  73292.000000   
unique                                        89128        1           NaN   
top     Remote SSH version: b'SSH-2.0-libssh-0.6.3'      ssh           NaN   
freq                                          38948    71172           NaN   
mean                                            NaN      NaN  43037.859343   
std                                             NaN      NaN  13565.286210   
min                                             NaN      NaN      0.000000   
25%                                             NaN      NaN  36112.750000   
50%                                             NaN      NaN  44834.500000   
75%                                             NaN      NaN  53400.000000   
max                                             NaN      NaN  65520.000000   

                   sensor  ...     type outfile  destfile duplicate  \
count              379033  ...      119       0         0     12201   
unique                  5  ...        1       0         0         1   
top     ubuntu_basic_pool  ...  ssh-rsa     NaN       NaN      True   
freq               304581  ...      119     NaN       NaN     12201   
mean                  NaN  ...      NaN     NaN       NaN       NaN   
std                   NaN  ...      NaN     NaN       NaN       NaN   
min  

ta bort dst_ip_identifier? 0 värden
timestamp --> alla är unika, inga olika tider


In [5]:
for col in df.columns:
    print(col)

session_id
dst_host_identifier
eventid
timestamp
src_ip_identifier
dst_ip_identifier
message
protocol
src_port
sensor
geolocation_data
arch
duration
ssh_client_version
username
password
hasshAlgorithms
macCS
langCS
compCS
encCS
hassh
kexAlgs
keyAlgs
fingerprint
key
type
outfile
destfile
duplicate
shasum
url
ttylog
size
filename
data


In [6]:
print(df['duration'].head)

<bound method NDFrame.head of 0              NaN
1              NaN
2              NaN
3              NaN
4         1.123112
            ...   
379028         NaN
379029         NaN
379030         NaN
379031         NaN
379032    1.656798
Name: duration, Length: 379033, dtype: float64>


In [7]:
df.isna().sum()

session_id                  0
dst_host_identifier         0
eventid                     0
timestamp                   0
src_ip_identifier           0
dst_ip_identifier      379033
message                     0
protocol               307861
src_port               305741
sensor                      0
geolocation_data            0
arch                   379033
duration               299219
ssh_client_version     314428
username               293475
password               293594
hasshAlgorithms        317589
macCS                  317589
langCS                 317589
compCS                 317589
encCS                  317589
hassh                  317589
kexAlgs                317589
keyAlgs                317589
fingerprint            378914
key                    378914
type                   378914
outfile                379033
destfile               379033
duplicate              366832
shasum                 366832
url                    379033
ttylog                 366832
size      

eventid, timestamp, message, sensor, src_ip_identifier --> inga nan värden
dst_ip_identifier, arch, filename, data --> ta bort

In [8]:
# Remove columns with too many missing values
# <30.000 missing values and irrelevant columns for the analysis 

df = df.drop(columns=["dst_ip_identifier", "arch", "hasshAlgorithms", "macCS", "langCS", "compCS", 
    "encCS", "hassh", "kexAlgs", "keyAlgs", "fingerprint", "key", "type",
    "outfile", "destfile", "duplicate", "shasum", "url", "ttylog", "size",
    "filename", "data"])

In [9]:
print(df.columns.tolist())
print(df.shape)

['session_id', 'dst_host_identifier', 'eventid', 'timestamp', 'src_ip_identifier', 'message', 'protocol', 'src_port', 'sensor', 'geolocation_data', 'duration', 'ssh_client_version', 'username', 'password']
(379033, 14)


(Can machine learning models accurately classify attackers into distinct behavioral profiles (automated bots, script kiddies, skilled operators) based on their command sequences and interaction patterns in SSH honeypots?)

Feature Extraction 

Temporal features: inter-command timing, session duration, time-of-day patterns


Command-based features: unique command ratio, command diversity, tool signatures


Behavioral patterns: reconnaissance vs. exploitation ratio, error rate, command correction attempts

In [10]:
df['timestamp'] = pd.to_datetime(df['timestamp'], errors='coerce')
df = df.sort_values(['session_id', 'timestamp'])

df.head()

session_id                                dst_host_identifier  \
173403  000122f2cea1  7e3a180ecc937536d7d7a0f2d3539cca6e28be9834d355...   
173404  000122f2cea1  7e3a180ecc937536d7d7a0f2d3539cca6e28be9834d355...   
173405  000122f2cea1  7e3a180ecc937536d7d7a0f2d3539cca6e28be9834d355...   
173406  000122f2cea1  7e3a180ecc937536d7d7a0f2d3539cca6e28be9834d355...   
173407  000122f2cea1  7e3a180ecc937536d7d7a0f2d3539cca6e28be9834d355...   

                       eventid                        timestamp  \
173403  cowrie.session.connect 2020-02-29 11:03:06.852680+00:00   
173404   cowrie.client.version 2020-02-29 11:03:06.856860+00:00   
173405       cowrie.client.kex 2020-02-29 11:03:07.106306+00:00   
173406     cowrie.login.failed 2020-02-29 11:03:08.444251+00:00   
173407     cowrie.login.failed 2020-02-29 11:03:10.914336+00:00   

                                        src_ip_identifier  \
173403  374d2a0a69375cd160fba8a3acd47c6b88c8f819f65405...   
173404  374d2a0a69375cd160fba8a3acd47c6b88c8f819f65405...   
173405  374d2a0a69375cd160fba8a3acd47c6b88c8f819f65405...   
173406  374d2a0a69375cd160fba8a3acd47c6b88c8f819f65405...   
173407  374d2a0a69375cd160fba8a3acd47c6b88c8f819f65405...   

                                                  message protocol  src_port  \
173403  New connection: 374d2a0a69375cd160fba8a3acd47c...      ssh   28464.0   
173404               Remote SSH version: b'SSH-2.0-PUTTY'     None       NaN   
173405  SSH client hassh fingerprint: 1616c6d18e845e7a...     None       NaN   
173406            login attempt [root/12345678910] failed     None       NaN   
173407             login attempt [root/12345678gf] failed     None       NaN   

                   sensor                                   geolocation_data  \
173403  ubuntu_basic_pool  {'continent_code': 'AS', 'country_code3': 'CN'...   
173404  ubuntu_basic_pool  {'continent_code': 'AS', 'country_code3': 'CN'...   
173405  ubuntu_basic_pool  {'continent_code': 'AS', 'country_code3': 'CN'...   
173406  ubuntu_basic_pool  {'continent_code': 'AS', 'country_code3': 'CN'...   
173407  ubuntu_basic_pool  {'continent_code': 'AS', 'country_code3': 'CN'...   

        duration ssh_client_version username     password  
173403       NaN               None     None         None  
173404       NaN   b'SSH-2.0-PUTTY'     None         None  
173405       NaN               None     None         None  
173406       NaN               None     root  12345678910  
173407       NaN               None     root   12345678gf

In [11]:
df['eventid'].value_counts().head(20)


eventid
cowrie.login.failed            80970
cowrie.session.connect         71172
cowrie.session.closed          67613
cowrie.client.version          64605
cowrie.client.kex              61444
cowrie.command.input           14320
cowrie.log.closed              12201
cowrie.login.success            4469
cowrie.direct-tcpip.request     2120
cowrie.client.fingerprint        119
Name: count, dtype: int64

In [12]:
def extract_command(row):
    msg = row["message"]

    # Only extract for command events
    if row["eventid"] != "cowrie.command.input":
        return np.nan
    
    if not isinstance(msg, str):
        return np.nan
    
    m = msg.lower()

    # Pattern 1: "CMD: <command>"
    if "cmd:" in m:
        return m.split("cmd:")[-1].strip()
    
    # Pattern 2: "input: <command>"
    if "input:" in m:
        return m.split("input:")[-1].strip()
    
    # Pattern 3: bare command (common)
    return m.strip()

df["command"] = df.apply(extract_command, axis=1)


In [13]:
# Inter-command timing
df["inter_command_time"] = df.groupby("session_id")["timestamp"].diff().dt.total_seconds()

# Session duration (use your existing duration + fallback)
session_times = df.groupby("session_id")["timestamp"].agg(["min", "max"])
session_times["session_duration"] = (session_times["max"] - session_times["min"]).dt.total_seconds()

# Time-of-day
df["hour"] = df["timestamp"].dt.hour
hour_features = df.groupby("session_id")["hour"].agg(
    mean_hour="mean",
    std_hour="std"
)


In [14]:
# #how many commands they typed
# how diverse those commands are
# how “human-like” or “bot-like” their command patterns are

command_features = df.groupby("session_id")["command"].agg(
    total_commands=lambda x: x.notna().sum(),
    unique_commands=lambda x: x.dropna().nunique()
)

command_features["unique_ratio"] = (
    command_features["unique_commands"] /
    command_features["total_commands"].replace(0, np.nan)
)


In [15]:
print(command_features.describe())

       total_commands  unique_commands  unique_ratio
count    71172.000000     71172.000000    858.000000
mean         0.201203         0.200739      0.998648
std          1.919671         1.912155      0.018396
min          0.000000         0.000000      0.692308
25%          0.000000         0.000000      1.000000
50%          0.000000         0.000000      1.000000
75%          0.000000         0.000000      1.000000
max         52.000000        36.000000      1.000000


In [16]:
tools = {
    "wget": "wget",
    "curl": "curl",
    "tftp": "tftp",
    "busybox": "busybox",
    "chmod": "chmod",
    "./": "exec_binary"
}

for tool, pattern in tools.items():
    df[f"tool_{tool}"] = df["message"].str.contains(pattern, case=False, na=False)

tool_features = df.groupby("session_id")[ [f"tool_{t}" for t in tools] ].sum()


In [17]:
print(df['tool_wget'].sum())
print(df['tool_curl'].sum())
print(df['tool_tftp'].sum())
print(df['tool_busybox'].sum()) 
print(df['tool_chmod'].sum())
print(df['tool_./'].sum())      

119
61
65
0
15
0


In [18]:
df.columns.tolist()


['session_id',
 'dst_host_identifier',
 'eventid',
 'timestamp',
 'src_ip_identifier',
 'message',
 'protocol',
 'src_port',
 'sensor',
 'geolocation_data',
 'duration',
 'ssh_client_version',
 'username',
 'password',
 'command',
 'inter_command_time',
 'hour',
 'tool_wget',
 'tool_curl',
 'tool_tftp',
 'tool_busybox',
 'tool_chmod',
 'tool_./']

In [19]:
# Combine all features into final per-session matrix
features = (
    session_times[['session_duration']]
    .join(hour_features, how='left')
    .join(command_features, how='left')
    .join(tool_features, how='left')
)

# Replace NaN with 0 for all non-timing numeric features
features = features.fillna(0)

print(features.head())
print(features.shape)
print(features.columns.tolist())


              session_duration  mean_hour  std_hour  total_commands  \
session_id                                                            
000122f2cea1          6.566378       11.0       0.0               0   
0001bec092c0          6.116296       14.0       0.0               0   
0001cb5ddeaf          2.578426       10.0       0.0               0   
00037bb3edc5          2.783038       23.0       0.0               0   
00060c49f409          3.156416        4.0       0.0               0   

              unique_commands  unique_ratio  tool_wget  tool_curl  tool_tftp  \
session_id                                                                     
000122f2cea1                0           0.0          0          0          0   
0001bec092c0                0           0.0          0          0          0   
0001cb5ddeaf                0           0.0          0          0          0   
00037bb3edc5                0           0.0          0          0          0   
00060c49f409          

In [20]:
features.isna().sum()


session_duration    0
mean_hour           0
std_hour            0
total_commands      0
unique_commands     0
unique_ratio        0
tool_wget           0
tool_curl           0
tool_tftp           0
tool_busybox        0
tool_chmod          0
tool_./             0
dtype: int64

In [21]:
features.describe()

session_duration     mean_hour      std_hour  total_commands  \
count      71172.000000  71172.000000  71172.000000    71172.000000   
mean          11.098681     11.622241      0.003023        0.201203   
std           32.964478      7.147578      0.136665        1.919671   
min            0.000000      0.000000      0.000000        0.000000   
25%            1.895061      5.000000      0.000000        0.000000   
50%            2.934816     12.000000      0.000000        0.000000   
75%            5.656439     18.000000      0.000000        0.000000   
max         1065.288987     23.000000     16.263456       52.000000   

       unique_commands  unique_ratio     tool_wget     tool_curl  \
count     71172.000000  71172.000000  71172.000000  71172.000000   
mean          0.200739      0.012039      0.001672      0.000857   
std           1.912155      0.109005      0.056699      0.029264   
min           0.000000      0.000000      0.000000      0.000000   
25%           0.000000      0.000000      0.000000      0.000000   
50%           0.000000      0.000000      0.000000      0.000000   
75%           0.000000      0.000000      0.000000      0.000000   
max          36.000000      1.000000      2.000000      1.000000   

          tool_tftp  tool_busybox    tool_chmod  tool_./  
count  71172.000000       71172.0  71172.000000  71172.0  
mean       0.000913           0.0      0.000211      0.0  
std        0.030207           0.0      0.014516      0.0  
min        0.000000           0.0      0.000000      0.0  
25%        0.000000           0.0      0.000000      0.0  
50%        0.000000           0.0      0.000000      0.0  
75%        0.000000           0.0      0.000000      0.0  
max        1.000000           0.0      1.000000      0.0

In [22]:
# Recon command patterns
recon_cmds = ["ls", "pwd", "whoami", "cat", "uname", "id", "ps"]

df["is_recon"] = df["command"].astype(str).str.contains(
    "|".join(recon_cmds),
    case=False,
    na=False
)

recon_features = (
    df.groupby("session_id")["is_recon"]
    .sum()
    .to_frame("recon_count")
)


In [23]:
exploit_cmds = ["wget", "curl", "tftp", "chmod", "busybox", "./"]

df["is_exploit"] = df["command"].astype(str).str.contains(
    "|".join(exploit_cmds),
    case=False,
    na=False
)

exploit_features = (
    df.groupby("session_id")["is_exploit"]
    .sum()
    .to_frame("exploit_count")
)

In [24]:
behavioral_ratio = recon_features.join(exploit_features, how="left")
behavioral_ratio["recon_exploit_ratio"] = (
    behavioral_ratio["recon_count"] /
    behavioral_ratio["exploit_count"].replace(0, np.nan)
)
behavioral_ratio = behavioral_ratio.fillna(0)


In [25]:
print (behavioral_ratio.sum())

recon_count            6857.0
exploit_count          6061.0
recon_exploit_ratio     954.7
dtype: float64


In [26]:
error_patterns = r"permission denied|not found|invalid|error|failed|unknown"

df["is_error"] = df["message"].str.contains(
    error_patterns,
    case=False,
    na=False
)

error_features = (
    df.groupby("session_id")["is_error"]
    .sum()
    .to_frame("error_count")
)


In [27]:
correction_patterns = r"\x08|backspace|command not found|invalid|not found"

df["is_correction"] = df["message"].str.contains(
    correction_patterns,
    case=False,
    na=False
)

correction_features = (
    df.groupby("session_id")["is_correction"]
    .sum()
    .to_frame("correction_count")
)


In [28]:
behavioral_features = (
    behavioral_ratio
    .join(error_features, how="left")
    .join(correction_features, how="left")
)

# Replace missing values with zero
behavioral_features = behavioral_features.fillna(0)

behavioral_features.head()


recon_count  exploit_count  recon_exploit_ratio  error_count  \
session_id                                                                   
000122f2cea1            0              0                  0.0            3   
0001bec092c0            0              0                  0.0            3   
0001cb5ddeaf            0              0                  0.0            1   
00037bb3edc5            0              0                  0.0            1   
00060c49f409            0              0                  0.0            1   

              correction_count  
session_id                      
000122f2cea1                 0  
0001bec092c0                 0  
0001cb5ddeaf                 0  
00037bb3edc5                 0  
00060c49f409                 0

In [29]:
print(behavioral_features.columns.tolist())


['recon_count', 'exploit_count', 'recon_exploit_ratio', 'error_count', 'correction_count']


In [30]:
behavioral_features.describe()


recon_count  exploit_count  recon_exploit_ratio   error_count  \
count  71172.000000   71172.000000         71172.000000  71172.000000   
mean       0.096344       0.085160             0.013414      1.137666   
std        0.917180       0.809597             0.125983      1.046012   
min        0.000000       0.000000             0.000000      0.000000   
25%        0.000000       0.000000             0.000000      1.000000   
50%        0.000000       0.000000             0.000000      1.000000   
75%        0.000000       0.000000             0.000000      1.000000   
max       14.000000      13.000000             3.000000     21.000000   

       correction_count  
count           71172.0  
mean                0.0  
std                 0.0  
min                 0.0  
25%                 0.0  
50%                 0.0  
75%                 0.0  
max                 0.0

In [31]:
features = (
    session_times[["session_duration"]]
    .join(hour_features, how="left")
    .join(command_features, how="left")
    .join(tool_features, how="left")
    .join(behavioral_features, how="left")
)

features = features.fillna(0)
features.head()


session_duration  mean_hour  std_hour  total_commands  \
session_id                                                            
000122f2cea1          6.566378       11.0       0.0               0   
0001bec092c0          6.116296       14.0       0.0               0   
0001cb5ddeaf          2.578426       10.0       0.0               0   
00037bb3edc5          2.783038       23.0       0.0               0   
00060c49f409          3.156416        4.0       0.0               0   

              unique_commands  unique_ratio  tool_wget  tool_curl  tool_tftp  \
session_id                                                                     
000122f2cea1                0           0.0          0          0          0   
0001bec092c0                0           0.0          0          0          0   
0001cb5ddeaf                0           0.0          0          0          0   
00037bb3edc5                0           0.0          0          0          0   
00060c49f409                0           0.0          0          0          0   

              tool_busybox  tool_chmod  tool_./  recon_count  exploit_count  \
session_id                                                                    
000122f2cea1             0           0        0            0              0   
0001bec092c0             0           0        0            0              0   
0001cb5ddeaf             0           0        0            0              0   
00037bb3edc5             0           0        0            0              0   
00060c49f409             0           0        0            0              0   

              recon_exploit_ratio  error_count  correction_count  
session_id                                                        
000122f2cea1                  0.0            3                 0  
0001bec092c0                  0.0            3                 0  
0001cb5ddeaf                  0.0            1                 0  
00037bb3edc5                  0.0            1                 0  
00060c49f409                  0.0            1                 0

In [32]:
timing_stats = df.groupby("session_id")["inter_command_time"].agg(
    mean_inter_command_time="mean",
    median_inter_command_time="median",
    std_inter_command_time="std",
    min_inter_command_time="min",
    max_inter_command_time="max",
).fillna(0) # Fill NaN with 0 for sessions with only one event (no time difference)

# Re-join all features 
features = features.join(timing_stats, how='left').fillna(0)


In [33]:
features.columns.tolist()

['session_duration',
 'mean_hour',
 'std_hour',
 'total_commands',
 'unique_commands',
 'unique_ratio',
 'tool_wget',
 'tool_curl',
 'tool_tftp',
 'tool_busybox',
 'tool_chmod',
 'tool_./',
 'recon_count',
 'exploit_count',
 'recon_exploit_ratio',
 'error_count',
 'correction_count',
 'mean_inter_command_time',
 'median_inter_command_time',
 'std_inter_command_time',
 'min_inter_command_time',
 'max_inter_command_time']

In [34]:
behavior_features = [
 'session_duration',
 'mean_hour',
 'std_hour',
 'total_commands',
 'unique_commands',
 'unique_ratio',
 'tool_wget',
 'tool_curl',
 'tool_tftp',
 'tool_busybox',
 'tool_chmod',
 'tool_./',
 'recon_count',
 'exploit_count',
 'recon_exploit_ratio',
 'error_count',
 'correction_count',
 'mean_inter_command_time',
 'median_inter_command_time',
 'std_inter_command_time',
 'min_inter_command_time',
 'max_inter_command_time'
]

X_cluster = features[behavior_features]

scaler = StandardScaler()
X_scaled = scaler.fit_transform(X_cluster)

kmeans = KMeans(n_clusters=3, random_state=42)
features["cluster"] = kmeans.fit_predict(X_scaled)

features["cluster"].value_counts()



cluster
0    68160
2     2997
1       15
Name: count, dtype: int64

In [ ]:
import hdbscan

clusterer = hdbscan.HDBSCAN(
    min_cluster_size=20,
    min_samples=10,
    metric='euclidean'
)

features["hdbscan_cluster"] = clusterer.fit_predict(X_scaled)

features["hdbscan_cluster"].value_counts()


In [ ]:
cluster_stats_mean = features.groupby("cluster").mean()
cluster_stats_median = features.groupby("cluster").median()
cluster_stats_min = features.groupby("cluster").min()
cluster_stats_max = features.groupby("cluster").max()

print("==== MEAN ====")
display(cluster_stats_mean.T)

print("==== MEDIAN ====")
display(cluster_stats_median.T)

print("==== MIN ====")
display(cluster_stats_min.T)

print("==== MAX ====")
display(cluster_stats_max.T)


==== MEAN ====


cluster                            0          1           2
session_duration            5.602010  97.229645  135.676966
mean_hour                  11.715103  13.988462    9.498473
std_hour                    0.001448   0.025467    0.038740
total_commands              0.185769  15.466667    0.475809
unique_commands             0.185769  13.533333    0.474474
unique_ratio                0.010182   0.943197    0.049614
tool_wget                   0.000015   0.666667    0.036036
tool_curl                   0.000000   0.533333    0.017684
tool_tftp                   0.000954   0.000000    0.000000
tool_busybox                0.000000   0.000000    0.000000
tool_chmod                  0.000000   1.000000    0.000000
tool_./                     0.000000   0.000000    0.000000
recon_count                 0.088806   6.533333    0.235569
exploit_count               0.078125   6.533333    0.212880
recon_exploit_ratio         0.011233   0.596667    0.060102
error_count                 1.187940   0.000000    0.000000
correction_count            0.000000   0.000000    0.000000
mean_inter_command_time     1.304135   2.461975  113.105012
median_inter_command_time   0.770238   0.734740  113.608687
std_inter_command_time      1.616078   9.898896    5.911547
min_inter_command_time      0.153615   0.009929  109.118091
max_inter_command_time      3.410427  66.005243  124.262447

==== MEDIAN ====


cluster                            0          1           2
session_duration            2.857803   6.041106  120.138033
mean_hour                  12.000000  14.000000    8.000000
std_hour                    0.000000   0.000000    0.000000
total_commands              0.000000  23.000000    0.000000
unique_commands             0.000000  16.000000    0.000000
unique_ratio                0.000000   0.960000    0.000000
tool_wget                   0.000000   1.000000    0.000000
tool_curl                   0.000000   1.000000    0.000000
tool_tftp                   0.000000   0.000000    0.000000
tool_busybox                0.000000   0.000000    0.000000
tool_chmod                  0.000000   1.000000    0.000000
tool_./                     0.000000   0.000000    0.000000
recon_count                 0.000000   6.000000    0.000000
exploit_count               0.000000   5.000000    0.000000
recon_exploit_ratio         0.000000   1.000000    0.000000
error_count                 1.000000   0.000000    0.000000
correction_count            0.000000   0.000000    0.000000
mean_inter_command_time     0.709224   0.694651  120.136390
median_inter_command_time   0.603963   0.710721  120.136390
std_inter_command_time      0.535218   0.652577    0.000000
min_inter_command_time      0.104472   0.003085  120.136390
max_inter_command_time      1.236544   1.592087  120.137015

==== MIN ====


cluster                      0          1          2
session_duration           0.0   3.956232  65.534272
mean_hour                  0.0  10.000000   0.000000
std_hour                   0.0   0.000000   0.000000
total_commands             0.0   1.000000   0.000000
unique_commands            0.0   1.000000   0.000000
unique_ratio               0.0   0.692308   0.000000
tool_wget                  0.0   0.000000   0.000000
tool_curl                  0.0   0.000000   0.000000
tool_tftp                  0.0   0.000000   0.000000
tool_busybox               0.0   0.000000   0.000000
tool_chmod                 0.0   1.000000   0.000000
tool_./                    0.0   0.000000   0.000000
recon_count                0.0   0.000000   0.000000
exploit_count              0.0   1.000000   0.000000
recon_exploit_ratio        0.0   0.000000   0.000000
error_count                0.0   0.000000   0.000000
correction_count           0.0   0.000000   0.000000
mean_inter_command_time    0.0   0.118065   4.224139
median_inter_command_time  0.0   0.054986   0.229762
std_inter_command_time     0.0   0.197597   0.000000
min_inter_command_time     0.0   0.001432   0.001414
max_inter_command_time     0.0   1.304983  60.241480

==== MAX ====


cluster                             0           1            2
session_duration           177.821801  639.923196  1065.288987
mean_hour                   23.000000   20.826923    23.000000
std_hour                    12.597619    0.382005    16.263456
total_commands              19.000000   52.000000    19.000000
unique_commands             19.000000   36.000000    19.000000
unique_ratio                 1.000000    1.000000     1.000000
tool_wget                    1.000000    2.000000     2.000000
tool_curl                    0.000000    1.000000     1.000000
tool_tftp                    1.000000    0.000000     0.000000
tool_busybox                 0.000000    0.000000     0.000000
tool_chmod                   0.000000    1.000000     0.000000
tool_./                      0.000000    0.000000     0.000000
recon_count                 11.000000   14.000000    11.000000
exploit_count                8.000000   13.000000     8.000000
recon_exploit_ratio          3.000000    1.750000     3.000000
error_count                 21.000000    0.000000     0.000000
correction_count             0.000000    0.000000     0.000000
mean_inter_command_time     53.585153   12.547514   120.238367
median_inter_command_time   53.585153    3.043462   120.238367
std_inter_command_time      69.659732   84.684031   130.385911
min_inter_command_time      53.585153    0.037773   120.238367
max_inter_command_time     132.800264  605.437558   825.427361

Cluster 0 = BOT

session_duration ≈ 2–6 sek

unique_ratio ≈ 0.00–0.01

total_commands ≈ 0

errors ≈ 1+

mean_inter_command_time ≈ 1s

inga verktyg (wget/curl/chmod = 0)


Cluster 1 = SCRIPT KIDDIE

unique_ratio very high → 0.69–1.00

total_commands high → 23 median, 15 mean

wget/curl/chmod = 1

recon/exploit = 5–6

session_duration ≈ 6–97 sek (median till mean)


Cluster 2 = SKILLED OPERATOR

session_duration very high → 120–135 mean, 65–~1000 range

mean_inter_command_time EXTREME → 113s median

min ICT ≈ 109s

total commands very low (0–1 median)

no wget/curl/chmod